homeassistant/components/octoprint/config_flow.py

```diff
@@ -0,0 +1,203 @@
+            vol.Optional(CONF_PATH, default=path): cv.string,
+            vol.Optional(CONF_SSL, default=ssl): cv.boolean,
+        },
+        extra=vol.ALLOW_EXTRA,
+    )
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect."""
+
+    session = async_get_clientsession(hass)
+    octoprint = OctoprintClient(
+        data[CONF_HOST], session, data[CONF_PORT], data[CONF_SSL], data[CONF_PATH]
+    )
+    octoprint.set_api_key(data[CONF_API_KEY])
+
+    uuid = None
+
+    try:
+        discovery = await octoprint.get_discovery_info()
+        if discovery:
+            uuid = discovery.upnp_uuid
+    except ApiError as conn_err:
+        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+        raise CannotConnect from conn_err
+
+    # Return info that you want to store in the config entry.
+    return {"title": data[CONF_HOST], "uuid": uuid}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for OctoPrint."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    api_key_task = None
+    user_input: dict = None
+
+    def __init__(self):
+        """Handle a config flow for OctoPrint."""
+        self.discovery_schema = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is None:
+            data = self.discovery_schema or _schema_with_defaults()
+            return self.async_show_form(step_id="user", data_schema=data)
+
+        self.user_input = user_input
+        if CONF_API_KEY in user_input:
+            return await self.async_step_finish(user_input)
+
+        self.api_key_task = None
+        return await self.async_step_get_api_key(user_input)
+
+    async def async_step_get_api_key(self, user_input=None):
+        """Get an Application Api Key."""
+        if not self.api_key_task:
+            self.api_key_task = self.hass.async_create_task(self._async_get_auth_key())
+            return self.async_show_progress(
+                step_id="get_api_key", progress_action="get_api_key"
+            )
+
+        try:
+            await self.api_key_task
+        except ApiError as err:
+            _LOGGER.error("Failed to get an application key : %s", err)
+            return self.async_show_progress_done(next_step_id="auth_failed")
+
+        return self.async_show_progress_done(next_step_id="finish")
+
+    async def async_step_finish(self, user_input=None):
+        """Finish the configuration setup."""
+        user_input = user_input or self.user_input
+        error = None
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            error = "cannot_connect"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            error = "unknown"
+        else:
+            if info["uuid"]:
```

Comment:
What is the consequence of a missing uuid and not setting entry unique_id? Won't that allow duplicate entries? When would this happen?